README.md
[![Packagist.org](https://img.shields.io/packagist/v/wikimedia/aho-corasick.svg?style=flat)](https://packagist.org/packages/wikimedia/aho-corasick)

AhoCorasick
===========

AhoCorasick is a PHP implementation of the [Aho-Corasick][1] string search
algorithm, which is an efficient way of searching a body of text for multiple
search keywords.

Here is how you use it:

```php
use AhoCorasick\MultiStringMatcher;

$keywords = new MultiStringMatcher( array( 'ore', 'hell' ) );

$keywords->searchIn( 'She sells sea shells by the sea shore.' );
// Result: array( array( 15, 'hell' ), array( 34, 'ore' ) )

$keywords->searchIn( 'Say hello to more text. MultiStringMatcher objects are reusable!' );
// Result: array( array( 4, 'hell' ), array( 14, 'ore' ) )
```


Features
--------

The algorithm works by constructing a finite-state machine out of the set of
search keywords. The time it takes to construct the finite state machine is
proportional to the sum of the lengths of the search keywords. Once
constructed, the machine can locate all occurences of all search keywords in
any body of text in a single pass, making exactly one state transition per
input character.

The algorithm originates from ["Efficient string matching: an aid to bibliographic search"][paper] (CACM, Volume 18, Issue 6, June 1975) by Alfred V. Aho and Margaret J. Corasick.

See also the definition and reference implementation on [nist.gov][dads].


[paper]: https://doi.org/10.1145/360825.36085
[dads]: http://xlinux.nist.gov/dads/HTML/ahoCorasick.html

Contribute
----------

- Issue tracker: https://phabricator.wikimedia.org/tag/ahocorasick/
- Source code: https://gerrit.wikimedia.org/g/AhoCorasick


Support
-------

If you are having issues, [please let us know][2].


License
-------

The project is licensed under the Apache license.


[1]: https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_string_matching_algorithm
[2]: https://phabricator.wikimedia.org/maniphest/task/create/?projects=PHID-PROJ-hs5ausnvlfs4e3n5gmzg


HISTORY.md
## v2.0.0

* [BREAKING CHANGE] Raise claimed PHP need from 7.0+ to 7.4+ (James D. Forrester)
* [BREAKING CHANGE] Drop HHVM support (James D. Forrester)
* Fix for numeric strings (Miguel Xavier Penha Neto) [T301632](https://phabricator.wikimedia.org/T301632)

## v1.0.1

* docs: Fix `<pre>` compat with Doxygen markdown (Timo Tijhof)

## v1.0.0

* Initial commit (Ori Livneh)


tests/AhoCorasickTest.php
<?php
/**
 * AhoCorasick PHP Library
 *
 * A PHP implementation of the Aho-Corasick string matching algorithm.
 *
 * Alfred V. Aho and Margaret J. Corasick, "Efficient string matching:
 *  an aid to bibliographic search", CACM, 18(6):333-340, June 1975.
 *
 * @link http://xlinux.nist.gov/dads//HTML/ahoCorasick.html
 * @link https://en.wikipedia.org/wiki/Aho-Corasick_string_matching_algorithm
 *
 * Copyright (C) 2015 Ori Livneh <ori@wikimedia.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file
 * @author Ori Livneh <ori@wikimedia.org>
 */

namespace AhoCorasick\Test;

use AhoCorasick\MultiStringMatcher;
use AhoCorasick\MultiStringReplacer;

/**
 * @covers \AhoCorasick\MultiStringMatcher
 */
class AhoCorasickTest extends \PHPUnit\Framework\TestCase {

	/**
	 * Sort results of NaiveMultiStringMatcher or MultiStringMatcher.
	 *
	 * Helps us assert equivalence.
	 *
	 * @param array &$matches Results array.
	 */
	public function sortMatcherResults( &$matches ) {
		// Sort the results by match offset, then by match length,
		// then by search keyword.
		usort( $matches, static function ( $a, $b ) {
			return ( $a[0] - $b[0] )
				?: ( strlen( $a[1] ) - strlen( $b[1] ) )
				?: strcmp( $a[1], $b[1] );
		} );
	}

	public function matcherCaseProvider() {
		$testCases = [
			[
				'She sells sea shells by the sea shore.',
				[ 's', 'se', 'sea', 'ore', 'hell', 'eat' ]
			],
			[
				'She sells sea shells by the sea shore.',
				[ 's', 'ls', 'lls', 'hells', 'shell', 'she', 'he', 'h' ],
			],
			[
				'井の中の蛙大海を知らず。',
				[ 'の', '', '食', '小蓑' ],
			],
			[
				'Вдохновение — это умение приводить '
					. 'себя в рабочее состояние.',
				[ 'это умение приводить себя' ],
			],
			[
				"初しぐれ猿も小蓑をほしげ也\nはつしぐれさるもこみのをほしげなり",
				[ "しげ也\nはつし" ],
			],
			[
				" (╯°□°）╯︵ ┻━┻  ",
				[ '°□°' ],
			],
			[
				'',
				[ 'a' ],
			],
		];

		return $testCases;
	}

	/** @dataProvider matcherCaseProvider */
	public function testMultiStringMatcher( $inputText, $searchKeywords ) {
		$referenceMatcher = new NaiveMultiStringMatcher( $searchKeywords );
		$referenceResults = $referenceMatcher->searchIn( $inputText );
		$this->sortMatcherResults( $referenceResults );

		$actualMatcher = new MultiStringMatcher( $searchKeywords );
		$actualResults = $actualMatcher->searchIn( $inputText );
		$this->sortMatcherResults( $actualResults );

		$this->assertEquals( $referenceResults, $actualResults );
	}

	/**
	 * @covers \AhoCorasick\MultiStringMatcher::__construct
	 */
	public function testConstructEmpty() {
		$this->expectWarning();
		$matcher = new MultiStringMatcher( [] );
	}

	/** @covers \AhoCorasick\MultiStringMatcher::getKeywords */
	public function testGetKeywords() {
		$searchKeywords = [ 's', 'sea', 'の' ];
		$matcher = new MultiStringMatcher( $searchKeywords );

		$this->assertEquals( $searchKeywords, $matcher->getKeywords() );
	}

	public function replacerCaseProvider() {
		return [
			[
				'The quick brown fox jumps over the lazy dog.',
				[ 'brown' => 'orange', 'brown fox' => 'blue cat', 'brown fox jx' => 'x' ]
			],
			[
				"It's raining snakes and ladders here",
				[ "It's" => 'It is', 'snake' => 'cat', 'ladder' => 'dog', 'here' => 'out there' ],
			],
			[
				'Now is the time for all good men to come to the aid of the party',
				[ 'USA' => 'United States' ],
			],
			[
				"富士の風や扇にのせて江戸土産\n" .
					"ふじのかぜやおうぎにのせてえどみやげ",
				[ '江戸' => '東京' ],
			],
		];
	}

	/**
	 * @dataProvider replacerCaseProvider
	 * @covers \AhoCorasick\MultiStringReplacer
	 */
	public function testMultiStringReplacer( $inputText, $replacePairs ) {
		$replacer = new MultiStringReplacer( $replacePairs );
		$actual = $replacer->searchAndReplace( $inputText );
		$expected = strtr( $inputText, $replacePairs );
		$this->assertEquals( $expected, $actual );
	}
}


tests/NaiveMultiStringMatcher.php
<?php
/**
 * AhoCorasick PHP Library
 *
 * A PHP implementation of the Aho-Corasick string matching algorithm.
 *
 * Alfred V. Aho and Margaret J. Corasick, "Efficient string matching:
 *  an aid to bibliographic search", CACM, 18(6):333-340, June 1975.
 *
 * @link http://xlinux.nist.gov/dads//HTML/ahoCorasick.html
 * @link https://en.wikipedia.org/wiki/Aho-Corasick_string_matching_algorithm
 *
 * Copyright (C) 2015 Ori Livneh <ori@wikimedia.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file
 * @author Ori Livneh <ori@wikimedia.org>
 */

namespace AhoCorasick\Test;

use AhoCorasick\MultiStringMatcher;

/**
 * A multiple-substring search using a naive algorithm.
 * (Iterate through the entire body of input text for each search keyword.)
 */
class NaiveMultiStringMatcher extends MultiStringMatcher {

	/**
	 * @param string $text The text to search in.
	 * @return array
	 */
	public function searchIn( $text ) {
		$matches = [];
		foreach ( $this->searchKeywords as $keyword => $length ) {
			$offset = 0;
            $keyword = strval( $keyword );
			while ( true ) {
				$offset = strpos( $text, $keyword, $offset );
				if ( $offset === false ) {
					break;
				}
				$matches[] = [ $offset, $keyword ];
				$offset = $offset + $length;
			}
		}
		return $matches;
	}
}


bench/bench.php
<?php
require_once __DIR__ . '/../src/MultiStringMatcher.php';
require_once __DIR__ . '/../src/MultiStringReplacer.php';

use AhoCorasick\MultiStringReplacer;
use AhoCorasick\MultiStringMatcher;

if ( !file_exists( __DIR__ . '/23835-0.txt' ) ) {
	die( "Please download http://www.gutenberg.org/files/23835/23835-0.txt\n" );
}

if ( !file_exists( __DIR__ . '/ZhConversion.php' ) ) {
	die( "You need ZhConversion.php, from " .
		"https://github.com/wikimedia/mediawiki/blob/master/includes/ZhConversion.php\n" );
}

require_once __DIR__ . '/ZhConversion.php';

$text = file_get_contents( __DIR__ . '/23835-0.txt' );

$options = getopt( '', [ 'count:', 'input:', 'profile', 'fss', 'msr', 'strtr' ] );
$text = file_get_contents( isset( $options['input'] ) ? $options['input'] : 'SueiTangYanYi.txt' );
$loops = isset( $options['count'] ) ? intval( $options['count'] ) : 5;
if ( !isset( $options['fss'] ) && !isset( $options['msr'] ) && !isset( $options['strtr'] ) ) {
	$options['fss'] = true;
	$options['msr'] = true;
	$options['strtr'] = true;
}
$profile = false;
if ( isset( $options['profile'] ) ) {
	$profile = true;
	$options['msr'] = true;
	unset( $options['fss'] );
	unset( $options['strtr'] );
}

if ( isset( $options['msr'] ) ) {
	$replacer = new MultiStringReplacer( $zh2Hant );
	if ( $profile ) {
		xhprof_enable( XHPROF_FLAGS_CPU );
	}
	$startTime = microtime( true );
	for ( $i = 0; $i < $loops; $i++ ) {
		$replacer->searchAndReplace( $text );
	}
	$endTime = microtime( true );
	$wallTime = 1000 * ( ( $endTime - $startTime ) / $loops );
	printf( "%-'.40s %.2fms\n", 'MultiStringRepeater::searchAndReplace(): ', $wallTime );
	if ( $profile ) {
		$profile = xhprof_disable();
		foreach ( $profile as $func => $data ) {
			printf( "%s: %.2f\n", $func, $data['cpu'] / $data['ct'] );
		}
	}
}

if ( function_exists( 'fss_prep_replace' ) && isset( $options['fss'] ) ) {
	$fss = fss_prep_replace( $zh2Hant );
	$startTime = microtime( true );
	for ( $i = 0; $i < $loops; $i++ ) {
		fss_exec_replace( $fss, $text );
	}
	$endTime = microtime( true );
	$wallTime = 1000 * ( ( $endTime - $startTime ) / $loops );
	printf( "%-'.40s %.2fms\n", 'fss_exec_replace(): ', $wallTime );
}

if ( isset( $options['strtr'] ) ) {
	$startTime = microtime( true );
	for ( $i = 0; $i < $loops; $i++ ) {
		strtr( $text, $zh2Hant );
	}
	$endTime = microtime( true );
	$wallTime = 1000 * ( ( $endTime - $startTime ) / $loops );
	printf( "%-'.40s %.2fms\n", 'strtr(): ', $wallTime );
}


bench/check.php
<?php
require_once __DIR__ . '/../src/MultiStringMatcher.php';
require_once __DIR__ . '/../src/MultiStringReplacer.php';

use AhoCorasick\MultiStringReplacer;
use AhoCorasick\MultiStringMatcher;

if ( !file_exists( __DIR__ . '/23835-0.txt' ) ) {
	die( 'Please download http://www.gutenberg.org/files/23835/23835-0.txt' );
}

if ( !file_exists( __DIR__ . '/ZhConversion.php' ) ) {
	die( 'You need ZhConversion.php, from http://git.io/vIMst' );
}

require_once __DIR__ . '/ZhConversion.php';

$text = file_get_contents( __DIR__ . '/23835-0.txt' );

$status = 0;
$expected = strtr( $text, $zh2Hant );

echo "MultiStringReplacer::searchAndReplace(): ";
$replacer = new MultiStringReplacer( $zh2Hant );
if ( $replacer->searchAndReplace( $text ) !== $expected ) {
	echo "ERROR\n";
	$status = 1;
} else {
	echo "OK\n";
}

if ( function_exists( 'fss_exec_replace' ) ) {
	echo "fss_exec_replace(): ";
	$fss = fss_prep_replace( $zh2Hant );
	if ( fss_exec_replace( $fss, $text ) !== $expected ) {
		echo "ERROR\n";
		$status = 1;
	} else {
		echo "OK\n";
	}
}

exit( $status );


src/MultiStringReplacer.php
<?php
/**
 * Copyright 2015 Ori Livneh <ori@wikimedia.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file
 * @author Ori Livneh <ori@wikimedia.org>
 */

namespace AhoCorasick;

/**
 * Search and replace functionality.
 */
class MultiStringReplacer extends MultiStringMatcher {

	/** @var array Mapping of states to outputs. */
	protected $replacePairs = [];

	/**
	 * Constructor.
	 *
	 * @param array $replacePairs array of ( 'from' => 'to' ) replacement pairs.
	 */
	public function __construct( array $replacePairs ) {
		foreach ( $replacePairs as $keyword => $replacement ) {
			if ( $keyword !== '' ) {
				$this->replacePairs[$keyword] = $replacement;
			}
		}
		parent::__construct( array_keys( $this->replacePairs ) );
	}

	/**
	 * Search and replace a set of keywords in some text.
	 *
	 * @param string $text The string to search in.
	 * @return string The input text with replacements.
	 *
	 * @par Example:
	 * @code
	 *   $replacer = new MultiStringReplacer( array( 'csh' => 'sea shells' ) );
	 *   $replacer->searchAndReplace( 'She sells csh by the sea shore.' );
	 *   // result: 'She sells sea shells by the sea shore.'
	 * @endcode
	 */
	public function searchAndReplace( $text ) {
		$state = 0;
		$length = strlen( $text );
		$matches = [];
		for ( $i = 0; $i < $length; $i++ ) {
			$ch = $text[$i];
			$state = $this->nextState( $state, $ch );
			foreach ( $this->outputs[$state] as $match ) {
				$offset = $i - $this->searchKeywords[$match] + 1;
				$matches[$offset] = $match;
			}
		}
		ksort( $matches );

		$buf = '';
		$lastInsert = 0;
		foreach ( $matches as $offset => $match ) {
			if ( $offset >= $lastInsert ) {
				$buf .= substr( $text, $lastInsert, $offset - $lastInsert );
				$buf .= $this->replacePairs[$match];
				$lastInsert = $offset + $this->searchKeywords[$match];
			}
		}
		$buf .= substr( $text, $lastInsert );

		return $buf;
	}
}


src/DeterministicMultiStringMatcher.php
<?php
/**
 * AhoCorasick PHP Library
 *
 * Copyright (C) 2015 Ori Livneh <ori@wikimedia.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file
 * @author Ori Livneh <ori@wikimedia.org>
 */

namespace AhoCorasick;

use AhoCorasick\MultiStringMatcher;

/**
 * Represents a variant of the Aho-Corasick string matching algorithm
 * which uses a deterministic finite automaton.
 *
 * The time it takes to construct the finite state machine is
 * proportional to the sum of the lengths of the search keywords.
 * Once constructed, the machine can locate all occurences of all
 * search keywords in a body of text in a single pass, making exactly
 * one state transition per input character.
 */
class DeterministicMultiStringMatcher extends MultiStringMatcher {

	/** @var array[] Mapping of all possible state transitions. **/
	protected $stateMachine = null;

	/**
	 * Constructor.
	 *
	 * @param string[] $searchKeywords The set of keywords to be matched.
	 */
	public function __construct( array $searchKeywords ) {
		parent::__construct( $searchKeywords );
		$this->computeFiniteStateMachine();
	}


	/**
	 * Map the current state and input character to the next state.
	 *
	 * @param int $currentState The current state of the string-matching
	 *  automaton.
	 * @param string $inputChar The character the string-matching
	 *  automaton is currently processing.
	 * @return int The state the automaton should transition to.
	 */
	public function nextState( $state, $ch ) {
		return isset( $this->stateMachine[$state][$ch] ) ?
			$this->stateMachine[$state][$ch] : 0;
	}

	/**
	 * Construct the string-matching finite state machine.
	 *
	 * The machine will make one state transition per input symbol.
	 */
	protected function computeFiniteStateMachine() {
		for ( $r = 0; $r < $this->numStates; $r++ ) {
			foreach ( $this->searchChars as $ch ) {
				$state = $r;
				while ( $state !== 0 && !isset( $this->yesTransitions[$state][$ch] ) ) {
					$state = $this->noTransitions[$state];
				}
				if ( isset( $this->yesTransitions[$state][$ch] ) ) {
					$this->stateMachine[$r][$ch] = $this->yesTransitions[$state][$ch];
				} else {
					$this->stateMachine[$r][$ch] = isset( $this->noTransitions[$state] )
						? $this->noTransitions[$state] : 0;
				}
			}
		}
	}
}


src/MultiStringMatcher.php
<?php
/**
 * Copyright 2015 Ori Livneh <ori@wikimedia.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file
 * @author Ori Livneh <ori@wikimedia.org>
 */

namespace AhoCorasick;

/**
 * Represents a finite state machine that can find all occurrences
 * of a set of search keywords in a body of text.
 *
 * The time it takes to construct the finite state machine is
 * proportional to the sum of the lengths of the search keywords.
 * Once constructed, the machine can locate all occurences of all
 * search keywords in a body of text in a single pass, making exactly
 * one state transition per input character.
 *
 * This is an implementation of the Aho-Corasick string matching
 * algorithm.
 *
 * Alfred V. Aho and Margaret J. Corasick, "Efficient string matching:
 *  an aid to bibliographic search", CACM, 18(6):333-340, June 1975.
 *
 * @link http://xlinux.nist.gov/dads/HTML/ahoCorasick.html
 */
class MultiStringMatcher {

	/** @var string[] The set of keywords to be searched for. */
	protected $searchKeywords = [];

	/** @var int The number of possible states of the string-matching finite state machine. */
	protected $numStates = 1;

	/** @var array Mapping of states to outputs. */
	protected $outputs = [];

	/** @var array Mapping of failure transitions. */
	protected $noTransitions = [];

	/** @var array Mapping of success transitions. */
	protected $yesTransitions = [];

	/**
	 * Constructor.
	 *
	 * @param string[] $searchKeywords The set of keywords to be matched.
	 */
	public function __construct( array $searchKeywords ) {
		foreach ( $searchKeywords as $keyword ) {
			if ( $keyword !== '' ) {
				$this->searchKeywords[$keyword] = strlen( $keyword );
			}
		}

		if ( !$this->searchKeywords ) {
			trigger_error( __METHOD__ . ': The set of search keywords is empty.', E_USER_WARNING );
			// Unreachable 'return' when PHPUnit detects trigger_error
			return; // @codeCoverageIgnore
		}

		$this->computeYesTransitions();
		$this->computeNoTransitions();
	}

	/**
	 * Accessor for the search keywords.
	 *
	 * @return string[] Search keywords.
	 */
	public function getKeywords() {
		return array_keys( $this->searchKeywords );
	}

	/**
	 * Map the current state and input character to the next state.
	 *
	 * @param int $currentState The current state of the string-matching
	 *  automaton.
	 * @param string $inputChar The character the string-matching
	 *  automaton is currently processing.
	 * @return int The state the automaton should transition to.
	 */
	public function nextState( $currentState, $inputChar ) {
		$initialState = $currentState;
		while ( true ) {
			$transitions =& $this->yesTransitions[$currentState];
			if ( isset( $transitions[$inputChar] ) ) {
				$nextState = $transitions[$inputChar];
				// Avoid failure transitions next time.
				if ( $currentState !== $initialState ) {
					$this->yesTransitions[$initialState][$inputChar] = $nextState;
				}
				return $nextState;
			}
			if ( $currentState === 0 ) {
				return 0;
			}
			$currentState = $this->noTransitions[$currentState];
		}
		// Unreachable outside 'while'
	} // @codeCoverageIgnore

	/**
	 * Locate the search keywords in some text.
	 *
	 * @param string $text The string to search in.
	 * @return array[] An array of matches. Each match is a vector
	 *  containing an integer offset and the matched keyword.
	 *
	 * @par Example:
	 * @code
	 *   $keywords = new MultiStringMatcher( array( 'ore', 'hell' ) );
	 *   $keywords->searchIn( 'She sells sea shells by the sea shore.' );
	 *   // result: array( array( 15, 'hell' ), array( 34, 'ore' ) )
	 * @endcode
	 */
	public function searchIn( $text ) {
		if ( !$this->searchKeywords || $text === '' ) {
			return [];  // fast path
		}

		$state = 0;
		$results = [];
		$length = strlen( $text );

		for ( $i = 0; $i < $length; $i++ ) {
			$ch = $text[$i];
			$state = $this->nextState( $state, $ch );
			foreach ( $this->outputs[$state] as $match ) {
				$offset = $i - $this->searchKeywords[$match] + 1;
				$results[] = [ $offset, $match ];
			}
		}

		return $results;
	}

	/**
	 * Get the state transitions which the string-matching automaton
	 * shall make as it advances through input text.
	 *
	 * Constructs a directed tree with a root node which represents the
	 * initial state of the string-matching automaton and from which a
	 * path exists which spells out each search keyword.
	 */
	protected function computeYesTransitions() {
		$this->yesTransitions = [ [] ];
		$this->outputs = [ [] ];
		foreach ( $this->searchKeywords as $keyword => $length ) {
			$state = 0;
			for ( $i = 0; $i < $length; $i++ ) {
				$ch = strval( $keyword )[$i];
				if ( !empty( $this->yesTransitions[$state][$ch] ) ) {
					$state = $this->yesTransitions[$state][$ch];
				} else {
					$this->yesTransitions[$state][$ch] = $this->numStates;
					$this->yesTransitions[] = [];
					$this->outputs[] = [];
					$state = $this->numStates++;
				}
			}

			$this->outputs[$state][] = $keyword;
		}
	}

	/**
	 * Get the state transitions which the string-matching automaton
	 * shall make when a partial match proves false.
	 */
	protected function computeNoTransitions() {
		$queue = [];
		$this->noTransitions = [];

		foreach ( $this->yesTransitions[0] as $ch => $toState ) {
			$queue[] = $toState;
			$this->noTransitions[$toState] = 0;
		}

		while ( true ) {
			$fromState = array_shift( $queue );
			if ( $fromState === null ) {
				break;
			}
			foreach ( $this->yesTransitions[$fromState] as $ch => $toState ) {
				$queue[] = $toState;
				$state = $this->noTransitions[$fromState];

				while ( $state !== 0 && empty( $this->yesTransitions[$state][$ch] ) ) {
					$state = $this->noTransitions[$state];
				}

				if ( isset( $this->yesTransitions[$state][$ch] ) ) {
					$noState = $this->yesTransitions[$state][$ch];
				} else {
					$noState = 0;
				}

				$this->noTransitions[$toState] = $noState;
				$this->outputs[$toState] = array_merge(
					$this->outputs[$toState], $this->outputs[$noState] );
			}
		}
	}
}
